### PR TITLE
build: Support musl target

### PR DESCRIPTION
**Pull request by sameo**
Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

### CODE REVIEW

**Comment by sameo** [comment]
We are really close to being able to build with musl and get a fully static build.
I think it's worth verifying that we can always build for both glibc and musl, but only run our CI on glibc as this is the main one we want to support.
So our support story would basically be:
* Builds and fully passes CI for x86-glibc
* Builds for x86-musl, and we'll take any fix for it as long as it does not break our CI.

**Comment by sboeuf** [comment]
@sameo yes that sounds good to me. This will give the opportunity for our users to build CH statically. Have you tried the static build with musl? Or is it static by default as you build with musl?

**Comment by sameo** [comment]
> @sameo yes that sounds good to me. This will give the opportunity for our users to build CH statically. Have you tried the static build with musl? Or is it static by default as you build with musl?

The build is static when you select the `musl` target:
```Shell
$ file target/x86_64-unknown-linux-musl/debug/cloud-hypervisor
target/x86_64-unknown-linux-musl/debug/cloud-hypervisor: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), statically linked, BuildID[sha1]=3c65fce03b981dc411672bb8cd09be0d4ebfdae9, with debug_info, not stripped
```

**Comment by rbradford** [comment]
Almost certainly need different versions of secomp filter list for the two? Or use a union...

**Comment by rbradford** [comment]
Have you tried the integration tests?

**Comment by sameo** [comment]
Pushed a change for running the CI with musl, please DO NOT MERGE (I strongly suspect CI will fail, but still...)

**Comment by rbradford** [comment]
@sameo all green now - need to decide if we only support/test musl or both

File: scripts/dev_cli.sh [hunk 1]
```diff
@@ -172,12 +174,19 @@ cmd_help() {
...
 cmd_build() {
     build="debug"
+    libc="gnu"
 
     while [ $# -gt 0 ]; do
 	case "$1" in
             "-h"|"--help")  { cmd_help; exit 1;     } ;;
             "--debug")      { build="debug";      } ;;
             "--release")    { build="release";    } ;;
+	    "--libc")
```
Comment (sboeuf) [inline scripts/dev_cli.sh:184]: nit: indentation is wrong

File: scripts/dev_cli.sh [hunk 2]
```diff
@@ -219,22 +231,31 @@ cmd_tests() {
...
-            "--unit")         { unit=true;      } ;;
-            "--cargo")        { cargo=true;    } ;;
-	    "--integration")  { integration=true;    } ;;
-	    "--all")          { cargo=true; unit=true; integration=true;  } ;;
-            "--")             { shift; break;         } ;;
+            "-h"|"--help")           { cmd_help; exit 1;     } ;;
+            "--unit")                { unit=true;      } ;;
+            "--cargo")               { cargo=true;    } ;;
+            "--integration")         { integration=true;    } ;;
+	    "--libc")
```
Comment (sboeuf) [inline scripts/dev_cli.sh:242]: ditto about indentation.

File: build.rs [hunk 3]
```diff
@@ -19,4 +20,16 @@ fn main() {
         "cargo:rustc-env=BUILT_VERSION={}",
         String::from_utf8(git_out.stdout).unwrap()
     );
+
+    // Default release path for a given target triple.
```
Comment (sboeuf) [inline build.rs:24]: What does "triple" mean in this case?

**Comment by sameo** [comment]
@sboeuf @rbradford Please do not fix yet.
I am fixing our `cargo test` build for the `musl` target. Building openssl with musl is exotic...

**Comment by rbradford** [comment]
> @sboeuf @rbradford Please do not merge yet.
> I am fixing our `cargo test` build for the `musl` target. Building openssl with musl is exotic...

Save your energy. I don't think it's really necessary to run the unit tests with musl. And definitely not required to use musl in the integration test framework as what we care about is the musl binary.

**Comment by sameo** [comment]
@sboeuf @rbradford This is ready for review now.

File: Jenkinsfile [hunk 4]
```diff
@@ -32,6 +32,16 @@ pipeline{
 								checkout scm
 							}
 						}
+						stage ('Run unit tests for musl') {
+							steps {
+								sh "scripts/dev_cli.sh tests --unit --libc musl"
+							}
+						}
+						stage ('Run integration tests for musl') {
```
Comment (rbradford) [inline Jenkinsfile]: We should run the musl integration tests on the master branch not the master node.

**Pull request merged**